### PR TITLE
Refactor: Simplify t4c services and enhance code maintainability

### PR DESCRIPTION
**Pull request by dominik003**
This PR simplifies the logic within the t4c model, t4c repository and t4c instance services by moving it from components to services. This change is significant as it reduces the complexity of the components, making the code easier to maintain and read. 

**Main changes include:**

- Adoption of the BehaviourSubject workflow within services.
- Reduction of reliance on rxjs within components.
- Removal of complex constructs such as `switchMaps`.

Through these changes, we've made the code more maintainable and readable, enhancing overall code quality.

### CODE REVIEW

**Comment by dominik003** [comment]
Already a first comment from me, so I won't forget it: I should write a test for the t4c instance update case that checks if the password is updated correctly if a non-empty string is given, and does not change if an empty string is given.

File: frontend/src/app/projects/models/model-detail/t4c-model-wrapper/t4c-model-wrapper.component.ts [hunk 1]
```diff
@@ -32,21 +32,19 @@ export class T4cModelWrapperComponent implements OnInit, OnDestroy {
...
-          this.t4cModelService.getT4CModel(project.slug, model.slug, t4cModelId)
-        )
-      )
-      .subscribe((t4cModel) => {
-        this.t4cModelService._t4cModel.next(t4cModel);
-        this.breadCrumbsService.updatePlaceholder({ t4cModel });
-      });
+    ]).subscribe(([project, model, t4cModelId]) =>
+      this.t4cModelService.loadT4CModel(project.slug, model.slug, t4cModelId)
+    );
```
Comment (MoritzWeber0) [inline frontend/src/app/projects/models/model-detail/t4c-model-wrapper/t4c-model-wrapper.component.ts:37]: The `untilDestroyed(this)` got lost here.

File: frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts [hunk 2]
```diff
@@ -79,105 +61,118 @@ export class ManageT4CModelComponent implements OnInit, OnDestroy {
     private toastService: ToastService
   ) {}
 
+  public form = new FormGroup({
+    t4cInstanceId: new FormControl<number | null>(null, Validators.required),
+    t4cRepositoryId: new FormControl<number | null>(
```
Comment (MoritzWeber0) [inline frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts:66]: Short note to myself: We should align if we want to use `ID` or `Id`. What's your opinion?

File: frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts [hunk 3]
```diff
@@ -191,33 +186,32 @@ export class ManageT4CModelComponent implements OnInit, OnDestroy {
...
   }
 
-  ngOnDestroy(): void {
-    this.t4cModelService._t4cModel.next(undefined);
-  }
-
   unlinkT4CModel() {
     if (!window.confirm(`Do you really want to unlink this T4C model?`)) {
       return;
     }
```
Comment (MoritzWeber0) [inline frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts:192]: I don't like these windows, they look a bit outdated and don't nicely integrate into our UI. We should get rid of them. Can also be handled in another PR.

File: frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.ts [hunk 4]
```diff
@@ -32,15 +31,8 @@ export class EditT4CInstanceComponent implements OnInit, OnDestroy {
...
-  get instance() {
-    return this._instance.value;
-  }
-
-  _capella_versions = new BehaviorSubject<ToolVersion[]>([]);
-  get capella_versions() {
-    return this._capella_versions.value;
-  }
+  instanceId?: number;
+  capella_versions?: ToolVersion[];
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/edit-t4c-instance/edit-t4c-instance.component.ts:35]:
```suggestion
  capellaVersions?: ToolVersion[];
```

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 5]
```diff
@@ -5,76 +5,133 @@
...
+  private _repositories = new BehaviorSubject<
+    T4CServerRepository[] | undefined
+  >(undefined);
+
+  readonly repositories = this._repositories.asObservable();
 
-  urlFactory(instance_id: number): string {
+  backendURLFactory(instance_id: number): string {
     return `${environment.backend_url}/settings/modelsources/t4c/${instance_id}/repositories/`;
   }
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:31]: Why the rename? We used `urlFactory` in other parts of the application.

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 6]
```diff
@@ -5,76 +5,133 @@
...
+      .post<null>(
+        `${this.backendURLFactory(instance_id)}${repository_id}/recreate/`,
+        {}
+      )
+      .pipe(tap(() => this.loadRepositories(instance_id)));
+  }
+
+  deleteRepository(
+    instance_id: number,
+    repository_id: number
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:97]:
```suggestion
    instanceId: number,
    repositoryId: number
```
Please update all parameters in this service to camelCase.

**Review by MoritzWeber0 — APPROVED** [review]
Such a nice refactoring. I really like that t4c related services and components are now aligned with our usual style in the frontend. Thanks @dominik003 :)

File: frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts (hunk 2 above)
Comment (dominik003) [reply frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts:66]: I prefer to use `Id`, but that's more of a personal preference than any real reason. So it would be fine to use `ID` as well.

File: frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts (hunk 3 above)
Comment (dominik003) [reply frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts:194]: Created #792 for this

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts (hunk 5 above)
Comment (dominik003) [reply frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:31]: I am not sure why I renamed it like this, so I would revert it. However, when looking at it again I renamed this here to `instanceUrlFactory(...)` and added another `repositoryUrlFactory(...)`

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts (hunk 6 above)
Comment (dominik003) [reply frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:97]: I already did this while handling the previous comment since I also noticed it there 😄

File: frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts (hunk 2 above)
Comment (MoritzWeber0) [reply frontend/src/app/projects/models/model-source/t4c/manage-t4c-model/manage-t4c-model.component.ts:66]: Ok, then let's use `Id` in the future 👍

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 7]
```diff
@@ -5,76 +5,131 @@
...
-  get repositories(): T4CServerRepository[] {
-    return this._repositories.value;
+  private _repositories = new BehaviorSubject<
+    T4CServerRepository[] | undefined
+  >(undefined);
+
+  readonly repositories = this._repositories.asObservable();
+
+  instanceUrlFactory(instanceId: number): string {
+    return `${environment.backend_url}/settings/modelsources/t4c/${instanceId}/repositories/`;
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:30]:
```suggestion
    return `${environment.backend_url}/settings/modelsources/t4c/${instanceId}/repositories`;
```

Don't use a trailing `/` in the urlFactory. Two reasons: 
- When just using the urlFactory as URL in requests, the request is done with a trailing slash. This leads to a redirect in fastAPI and therefore another (unnecessary) request. 
- Concatenation looks weird, e.g., `${this.repositoryUrlFactory(instanceId, repositoryId)}start/` instead of  `${this.repositoryUrlFactory(instanceId, repositoryId)}/start` (the latter looks a lot better imho)

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 8]
```diff
@@ -5,76 +5,131 @@
...
   }
 
-  urlFactory(instance_id: number): string {
-    return `${environment.backend_url}/settings/modelsources/t4c/${instance_id}/repositories/`;
+  repositoryUrlFactory(instanceId: number, repositoryId: number): string {
+    return `${this.instanceUrlFactory(instanceId)}${repositoryId}/`;
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:34]:
```suggestion
    return `${this.instanceUrlFactory(instanceId)}${repositoryId}`;
```

Same as above.

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 9]
```diff
@@ -5,76 +5,131 @@
...
-    repository_id: number
-  ): Observable<null> {
-    return this.http.delete<null>(
-      `${this.urlFactory(instance_id)}${repository_id}/`
-    );
+  startRepository(instanceId: number, repositoryId: number): Observable<null> {
+    this.publishRepositoriesWithChangedStatus(repositoryId, 'LOADING');
+    return this.http
+      .post<null>(
+        `${this.repositoryUrlFactory(instanceId, repositoryId)}start/`,
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:67]:
```suggestion
        `${this.repositoryUrlFactory(instanceId, repositoryId)}/start`,
```

See above.

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 10]
```diff
@@ -5,76 +5,131 @@
...
+    repositoryId: number
   ): Observable<null> {
-    return this.http.post<null>(
-      `${this.urlFactory(instance_id)}${repository_id}/start/`,
-      {}
-    );
+    this.publishRepositoriesWithChangedStatus(repositoryId, 'LOADING');
+    return this.http
+      .post<null>(
+        `${this.repositoryUrlFactory(instanceId, repositoryId)}recreate/`,
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:90]:
```suggestion
        `${this.repositoryUrlFactory(instanceId, repositoryId)}/recreate`,
```

See above.

File: frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts [hunk 11]
```diff
@@ -5,76 +5,131 @@
...
-    return this._repositories.value;
+  private _repositories = new BehaviorSubject<
+    T4CServerRepository[] | undefined
+  >(undefined);
+
+  readonly repositories = this._repositories.asObservable();
+
+  instanceUrlFactory(instanceId: number): string {
+    return `${environment.backend_url}/settings/modelsources/t4c/${instanceId}/repositories/`;
   }
```
Comment (MoritzWeber0) [inline frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:31]: This belongs to the `T4CInstanceService` and can be called from the T4CRepoService.

Comment (dominik003) [reply frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts:31]: It does not really belong to the instance service as we have the `/repositories` at the end. Wasn't sure about the naming here so probably something like `baseUrlFactory` + `repositoryUrlFactory` would be better?

Comment (MoritzWeber0) [reply frontend/src/app/settings/modelsources/t4c-settings/service/t4c-repos/t4c-repo.service.ts]: In this case, I'd just just `${environment.backend_url}/settings/modelsources/t4c/${instanceId}` in the t4cInstanceService and then add `/repositories` in the individual HTTP requests. It's only used in `loadRepositories` and `createRepository`.
